File: src/shape.rs
use js_sys::Uint32Array;

pub fn get_size(shape: &Vec<usize>) -> usize {
    if shape.len() == 0 {
        return 0;
    }

    let mut size: usize = 1;
    for sh in shape {
        size *= sh;
    }
    return size;
}

pub fn get_size_from(shape: &Vec<usize>, start_ix: usize) -> usize {
    if shape.len() == 0 {
        return 0;
    }

    let mut size: usize = 1;
    for i in start_ix..shape.len() {
        size *= shape[i];
    }
    return size;
}

pub fn get_size_from_to(shape: &Vec<usize>, start_ix: usize, end_ix: usize) -> usize {
    if shape.len() == 0 {
        return 0;
    }

    let mut size: usize = 1;
    for i in start_ix..end_ix {
        size *= shape[i];
    }
    return size;
}

pub fn compute_strides(shape: &Vec<usize>) -> Vec<usize> {
    let rank = shape.len();

    if rank == 0 {
        return vec![];
    }
    if rank == 1 {
        if shape[0] == 1 {
            return vec![0];
        }
        return vec![1];
    }

    let mut res: Vec<usize> = vec![1; rank];
    if shape[rank - 1] == 1 {
        res[rank - 1] = 0;
    } else {
        res[rank - 1] = 1;
    }
    let mut last_stride = 1;

    for i in (0..(shape.len()-1)).rev() {
        last_stride = last_stride * shape[i+1];
        if shape[i] == 1 {
            res[i] = 0;
        } else {
            res[i] = last_stride;
        }
    }

    return res;
}

pub fn compute_strides_no_zero(shape: &Vec<usize>) -> Vec<usize> {
    let rank = shape.len();

    if rank == 0 {
        return vec![];
    }
    if rank == 1 {
        return vec![1];
    }

    let mut res: Vec<usize> = vec![1; rank];

    let mut last_stride = 1;

    for i in (0..(shape.len()-1)).rev() {
        last_stride = last_stride * shape[i+1];
        res[i] = last_stride;
    }

    return res;
}

pub fn compute_strides_uint32(shape: &Uint32Array) -> Vec<usize> {
    let rank = shape.length();

    if rank == 0 {
        return vec![];
    }
    if rank == 1 {
        if shape.get_index(0) == 0 {
            return vec![0];
        }
        return vec![1];
    }

    let mut res: Vec<usize> = vec![1; rank as usize];
    if shape.get_index(rank - 1) == 1 {
        res[rank as usize - 1] = 0;
    } else {
        res[rank as usize - 1] = 1;
    }
    let mut last_stride = 1;

    for i in (0..(shape.length()-1)).rev() {
        last_stride = last_stride * (shape.get_index(i + 1) as usize);
        if shape.get_index(i) == 1 {
            res[i as usize] = 0;
        } else {
            res[i as usize] = last_stride;
        }
    }

    return res;
}

pub fn compare_shapes(a: &Vec<usize>, b: &Vec<usize>) -> bool {
    if a.len() != b.len() {
        return false;
    }

    for i in 0..a.len() {
        if a[i] != b[i] {
            return false;
        }
    }

    return true;
}

pub fn index_to_pos(index: &Vec<usize>, strides: &Vec<usize>) -> usize {
    let mut ix: usize = 0;
    for i in 0..index.len() {
        ix += index[i] * strides[i];
    }
    return ix;
}

pub fn pos_to_index(pos: usize, strides: &Vec<usize>) -> Vec<usize> {
    let mut ix = vec![0; strides.len()];

    let mut pos_c = pos;

    for i in 0..strides.len() {
        ix[i] = pos_c / strides[i];
        pos_c %= strides[i];
    }
    return ix;
}

pub fn increment_index(index: &mut Vec<usize>, shape: &Vec<usize>) {
    for i in (0..index.len()).rev() {
        index[i] += 1;
        if index[i] >= shape[i] {
            index[i] = 0;
        } else {
            return;
        }
    }
}

pub fn decrement_index(index: &mut Vec<usize>, shape: &Vec<usize>) {
    for i in (0..index.len()).rev() {
        if index[i] == 0 {
            index[i] = shape[i] - 1;
        } else {
            index[i] -= 1;
            return;
        }
    }
}

File: lib/util/shape.ts
export function getSize(shape: ReadonlyArray<number>, zeroSize = 0) {
  if (shape.length === 0) {
    return zeroSize;
  }

  let size = 1;
  for (let i = 0; i < shape.length; i += 1) {
    size *= shape[i];
  }
  return size;
}

export function computeStrides(shape: ReadonlyArray<number>) {
  const rank = shape.length;

  if (rank === 0) {
    return [];
  }
  if (rank === 1) {
    if (shape[0] === 1) {
      return [0];
    } else {
      return [1];
    }
  }

  const strides = new Array(rank);
  strides[rank - 1] = 1;
  if (shape[rank - 1] === 1) {
    strides[rank - 1] = 0;
  }
  let lastStride = 1;
  for (let i = rank - 2; i >= 0; i -= 1) {
    lastStride = shape[i + 1] * lastStride;
    if (shape[i] === 1) {
      strides[i] = 0;
    } else {
      strides[i] = lastStride;
    }
  }

  return strides;
}

export function indexToPos(
  index: ReadonlyArray<number>,
  strides: ReadonlyArray<number>,
  shape?: ReadonlyArray<number>
) {
  let ix = 0;
  for (let i = 0; i < index.length; i += 1) {
    if (shape) {
      if (index[i] < 0 || (index[i] >= shape[i] && shape[i] !== 1)) {
        throw new Error('Invalid index');
      }
    }
    ix += index[i] * strides[i];
  }
  return ix;
}

export function posToIndex(
  pos: number,
  strides: ReadonlyArray<number>
): number[] {
  let res = pos;
  const rank = strides.length;
  const index = new Array(rank);

  for (let i = 0; i < index.length; i += 1) {
    index[i] = Math.floor(res / strides[i]);
    res %= strides[i];
  }

  return index;
}

export function compareShapes(
  a: ReadonlyArray<number>,
  b: ReadonlyArray<number>
) {
  if (a.length !== b.length) {
    return false;
  }

  for (let i = 0; i < a.length; i += 1) {
    if (a[i] !== b[i]) {
      return false;
    }
  }

  return true;
}

export function checkEquivShapes(
  a: ReadonlyArray<number>,
  b: ReadonlyArray<number>
) {
  if (a.length !== b.length) {
    return false;
  }

  for (let i = 0; i < a.length; i += 1) {
    if (a[i] !== b[i] && a[i] !== 1 && b[i] !== 1) {
      return false;
    }
  }

  return true;
}

export function incrementIndex(index: number[], shape: readonly number[]) {
  for (let i = index.length - 1; i >= 0; i--) {
    index[i] += 1;
    if (index[i] >= shape[i]) {
      index[i] = 0;
    } else {
      break;
    }
  }
}

export function decrementIndex(index: number[], shape: readonly number[]) {
  for (let i = index.length - 1; i >= 0; i--) {
    index[i] -= 1;
    if (index[i] < 0) {
      index[i] = shape[i] - 1;
    } else {
      break;
    }
  }
}
